docs(api): new_labware.rst revisions for Flex

jwwojak opened the pull request:
# Overview

Suggested text and code example revisions to the [Labware](https://docs.opentrons.com/v2/new_labware.html) API docs (`new_labware.rst` file) to include Flex code and code examples.

# Test Plan

- Checked links and cross-references to make sure they work.
- Code changes are minor. For example, adding new code samples to show labware in a Flex deck slot is just a change from `1` (OT-2) to `D1` (Flex).

# Changelog

- Added a new page introduction
- Added and revised text to define default and custom labware
- Style elements to improve content display.

Revised note: Will now include info about the `load_labware` `adapter` argument. See RLAB-343 and PR https://github.com/Opentrons/opentrons/pull/13016.

# Review requests

@ecormany 

# Risk assessment

Low risk. Code changes are minor, mostly deck slot coordinates and text.

jwwojak committed:
Update new_labware.rst

Starting changes to the tutorial docs (tutorial.rst). Incorporate Flex and Flex-related stuff into the docs.

jwwojak committed:
Update new_labware.rst

Starting save.

jwwojak committed:
Update new_labware.rst

Putting labware in a table. Similar to the tutorial.rst page.

jwwojak committed:
Update new_labware.rst

Intro section revisions.

jwwojak committed:
Update new_labware.rst

End of day commit.

jwwojak committed:
Merge branch 'edge' into labware-revisions

jwwojak committed:
Update new_labware.rst

jwwojak committed:
Merge branch 'edge' into labware-revisions

jwwojak committed:
Update new_labware.rst

Latest revisions. Saving and may open draft PR.

jwwojak committed:
Update new_labware.rst

Fixing syntax errors with links. Added link to deck_slots.rst doc.

jwwojak committed:
Update new_labware.rst

Final final, maybe?

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,76 @@
 Labware
 #######
 
+Labware are the consumable items that you work with, reuse, or discard while running a protocol on a Flex or OT-2. Items such as pipette tips, well plates, tubes, and reservoirs are all examples of labware. This section provides a brief overview of default labware, custom labware, and how to use basic labware API methods to create a protocol for your robot.
```
Since we're using the language from the manual here, I think "durable or consumable" is appropriate. The `Labware` class includes things like aluminum adapters, which you definitely shouldn't throw out!

ecormany replied:
Maybe also "when creating a protocol for your robot." Saying "to create" makes it sound like those are the only tools you need.

jwwojak replied:
Thanks, Ed. I had "durable" in here from the manual, but (as you noticed) removed it. Was probably over-thinking it to stay focused on the disposable items. Putting it back in. Will change text in 2nd sentence too.

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,76 @@
 
-Default Labware
-^^^^^^^^^^^^^^^
+From the example in the :ref:`overview-section-v2`, we used the :py:meth:`.ProtocolContext.load_labware` method to load Flex and OT-2 labware as shown below. 
```
```suggestion
Following the example in the :ref:`overview-section-v2`, here is how to the :py:meth:`.ProtocolContext.load_labware` method to load labware on either Flex or OT-2. 
```
Something like this, mostly to avoid the noun-noun compound "Flex and OT-2 labware"

ecormany replied:
oops, it would help if i didn't suggest typos 🤦‍♂️ 
"here is how to **use** the … method"

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,76 @@
+    plate = protocol.load_labware('corning_96_wellplate_360ul_flat', '2')
 
-If you have a piece of labware that is not in the Labware Library, you can create your own definition using the `Opentrons Labware Creator <https://labware.opentrons.com/create/>`_. Before using the Labware Creator, you should read the introduction article `here <https://support.opentrons.com/en/articles/3136504-creating-custom-labware-definitions>`__.
+After the ``load_labware`` method loads labware into your protocol, it returns a :py:class:`opentrons.protocol_api.labware.Labware` object.
```
"When" the method loads labware? I think this is considered to be simultaneous, or nearly so. But maybe check with a Python guru…

jwwojak replied:
An easy enough change to make, but does it really make a difference to an average user? There are 2 actions here:

1. Method loads labware
2. API returns an object

That this happens almost immediately isn't relevant to the purpose of the text and intended audience, though it might matter to an engineer. Visions of angels dancing on the head of a pin swim before my eyes. Will change to "when" nonetheless.

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,76 @@
+After the ``load_labware`` method loads labware into your protocol, it returns a :py:class:`opentrons.protocol_api.labware.Labware` object.
 
-Once you have created your labware and saved it as a ``.json`` file, you can add it to the Opentrons App by clicking "More" and then "Labware". Once you have added your labware to the Opentrons App, it will be available to all Python Protocol API version 2 protocols uploaded to your robot through that Opentrons App. If other people will be using this custom labware definition, they must also add  ...
+.. Recommend a call-out instead of a section (H2, H3) for "label." It's more a "pro tip" nice-to-have rather than
```
Yes, this is fine, the bit of content is so short. Maybe we should give it a label so it can be cross-referenced easily.

jwwojak replied:
added a label. Here's the whole call-out:

```
.. _labware-label:

.. Tip::
    
    The ``load_labware`` method includes an optional ``label`` argument. You can use it to identify labware with a descriptive name. If used, the label value is displayed in the Opentrons App. For example::
        
        tiprack = protocol.load_labware(
            load_name= 'corning_96_wellplate_360ul_flat',
            location= 'D1',  # Flex deck slot
            label= 'any-name-you-want')
```

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,76 @@
+Custom Labware
+^^^^^^^^^^^^^^
 
+Custom labware is labware that is not listed the Labware Library. If your protocol needs something that's not in the library, you can create it with the `Opentrons Labware Creator <https://labware.opentrons.com/create/>`_. However, before using the Labware Creator, you should take a moment to review the support article, `Creating Custom Labware Definitions <https://support.opentrons.com/s/article ...
```
```suggestion
Custom labware is labware that is not listed the Labware Library. If your protocol needs something that's not in the library, you can create it with the `Opentrons Labware Creator <https://labware.opentrons.com/create/>`_. However, before using the Labware Creator, you should take a moment to review the support article, `Creating Custom Labware Definitions <https://support.opentrons.com/s/article/Creating-Custom-Labware-Definitions>`_.
```
stray space

jwwojak replied:
oops!

jwwojak committed:
Update new_labware.rst

Trying to get the section symbols and headers sorted out.

Also, will add text for the load_labware `adapter` parameter.

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,76 @@
+    * - Well plate
+      - `Corning 96 Well Plate 360 µL Flat <https://labware.opentrons.com/corning_96_wellplate_360ul_flat/>`_
+      - ``corning_96_wellplate_360ul_flat``
+    * - Flex tiprack
```
"tip rack", with a space, in prose

jwwojak replied:
Change "tiprack" to "tip rack", except in code.

ecormany reviewed (changes requested):
Looking good so far! I haven't done a detailed read below Accessing Wells in Labware, but I get the sense that not much has changed there. Will do a top-to-bottom read when this is marked ready for review.

jwwojak committed:
Update new_labware.rst

Adding "durable" to 1st sentence:  "durable or consumable".

Change in last sentence from "to create" to "when creating a protocol..."

jwwojak committed:
Update api/docs/v2/new_labware.rst

Removing extra space before period.

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

jwwojak committed:
Update new_labware.rst

Change text from "tiprack" to "tip rack" except for code.

jwwojak committed:
Update api/docs/v2/new_labware.rst

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

jwwojak committed:
Update new_labware.rst

jwwojak committed:
Update new_labware.rst

Making changes from reviewer feedback. Starting the `adapter` section. Committing to save changes.

jwwojak committed:
Update new_labware.rst

Pre-meeting push. Working out labware-adapter text and organization.

jwwojak committed:
Update new_labware.rst

Fooling around with `:attr:` too much. Saving recent changes.

jwwojak committed:
Update new_labware.rst

Adding loading labware on adapters. Examples not tested, need to test. Saving.

jwwojak committed:
Update new_labware.rst

Spinning out here. Going to move on to the next section.

jwwojak committed:
Update new_labware.rst

Got a context error with my code example. Checked with Jeremy and this only has to do with me not using the latest build. Code is correct. Pushing and moving PR out of draft status.

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,110 @@
+.. Description almost verbatim from Flex user manual.
 
-In the example given in the :ref:`overview-section-v2` section, we loaded labware like this:
+Default labware is everything listed in the `Opentrons Labware Library <https://labware.opentrons.com/>`_. When used in a protocol, your Flex or OT-2 knows how to work with default labware. However, you must first inform the API about the labware you will place on the robot’s deck. Search the library when you’re looking for the API load names of the labware you want to use. You can copy the load  ...
```
Would be good to link the method here, but keep the clean, short name.
```suggestion
Default labware is everything listed in the `Opentrons Labware Library <https://labware.opentrons.com/>`_. When used in a protocol, your Flex or OT-2 knows how to work with default labware. However, you must first inform the API about the labware you will place on the robot’s deck. Search the library when you’re looking for the API load names of the labware you want to use. You can copy the load names from the library and pass them to the :py:meth:`~.ProtocolContext.load_labware` method in your protocol.
```

jwwojak replied:
Committing change.

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,110 @@
 
-    plate = protocol.load_labware('corning_96_wellplate_360ul_flat', '2')
-    tiprack = protocol.load_labware('opentrons_96_tiprack_300ul', '1')
+Custom labware is labware that is not listed the Labware Library. If your protocol needs something that's not in the library, you can create it with the `Opentrons Labware Creator <https://labware.opentrons.com/create/>`_. However, before using the Labware Creator, you should take a moment to review the support article, `Creating Custom Labware Definitions <https://support.opentrons.com/s/article ...
```
```suggestion
Custom labware is labware that is not listed the Labware Library. If your protocol needs something that's not in the library, you can create it with the `Opentrons Labware Creator <https://labware.opentrons.com/create/>`_. However, before using the Labware Creator, you should take a moment to review the support article `Creating Custom Labware Definitions <https://support.opentrons.com/s/article/Creating-Custom-Labware-Definitions>`_.
```
not an appositive

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,110 @@
 
-Default Labware
-^^^^^^^^^^^^^^^
+When the ``load_labware`` method loads labware into your protocol, it returns a :py:class:`opentrons.protocol_api.labware.Labware` object.
```
```suggestion
When the ``load_labware`` method loads labware into your protocol, it returns a :py:class:`~opentrons.protocol_api.labware.Labware` object.
```
tilde trick to shorten this cross-reference

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,110 @@
+        
+        tiprack = protocol.load_labware(
+            load_name= 'corning_96_wellplate_360ul_flat',
+            location= 'D1',  # Flex deck slot
```
can remove comment here, since slots are cross-compatible

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,110 @@
 
-Custom Labware
-^^^^^^^^^^^^^^
+The the previous section demonstrates loading labware directly into a deck slot. But, the ``load_labware`` method also includes the ``adapter`` argument. This parameter lets you stack labware on top of a compatible adapter and work with them as a single unit in a protocol. For example, you can use an adapter as a simple labware holding device or when you need to use a labware/adapter combination  ...
```
```suggestion
The previous section demonstrates loading labware directly into a deck slot. But, the ``load_labware`` method also includes the ``adapter`` argument. This parameter lets you stack labware on top of a compatible adapter and work with them as a single unit in a protocol. For example, you can use an adapter as a simple labware holding device or when you need to use a labware/adapter combination with a module. The ability to combine labware with adapters adds functionality and flexibility to your robot and protocols. 
```
☝️ typo fix

additionally, i think this section could shift in approach. loading modules in examples starts to overcomplicate things — especially since we can do this all directly on the deck now. i'd propose instead of…
- On the Deck
- On a Module
…to have subsections for…
- Using `load_labware`
- Using `load_adapter`

we definitely need to document `load_adapter()` and this seems like the place. load an adapter straight on the deck, then add labware on top. then cross reference to Modules page for more complex examples that do that on top of a module.

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,110 @@
+    
+    corning_plate = protocol.load_labware (
+        load_name= 'corning_96_wellplate_360ul_flat',
+        location= "D1", #Flex deck slot
```
```suggestion
        location= "D1",
```

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -6,51 +6,110 @@
+        load_name= 'corning_96_wellplate_360ul_flat',
+        location= "D1", #Flex deck slot
+        adapter= "opentrons_universal_flat_adapter")
+
```
include versionadded here too (the parameter is new)

jwwojak replied:
Added

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -76,23 +135,21 @@ The ending well will be in the bottom right, see the diagram below for further e
-    :substitutions:
-
-    '''
-    Examples in this section expect the following
```
Too much context removed…the code blocks are floating free below the image. How about:

```
The code in this section assumes that ``plate`` is a 24-well plate. For example:
<block loading the corning plate into D1>
```
then skip the OT-2 block that's only 1 character different.

jwwojak replied:
Fixed. Assuming "then skip the OT-2 block" means remove it. Makes sense. Revised note at top tells readers the samples use Flex deck slots and links to the deck slot doc. Will resolve and commit.

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -266,64 +320,60 @@ Well Dimensions
 
 The functions in the :ref:`new-well-access` section above return a single :py:class:`.Well` object or a larger object representing many wells. :py:class:`.Well` objects have attributes that provide information about their physical shape, such as the depth or diameter, as specified in their corresponding labware definition. These properties can be used for different applications, such as calculati ...
 
+.. note::
+
+    In the code samples below, the ``load_labware`` method shows a Flex deck slot location (``D1``). If you have an OT-2, replace ``D1`` with ``1``. Although our Python API knows how to translate Flex and OT-2 deck locations, it's a good practice to use the deck location coordinates that match your robot model. See :ref:`deck-slots` for more information.
+
```
we could move (or copy) this further up the page, wherever we do this first.

jwwojak replied:
Moved to top, under the intro, and shortened.

```
.. note::

Code snippets use alphanumeric Flex deck slot locations (e.g. ``D1``, ``D2``, etc.). 
If you have an OT-2, replace the Flex deck slot coordinate with its OT-2 equivalent.
For example, Flex slot D1 corresponds to slot 1 on an OT-2. See :ref:`deck-slots` for more information.
```

jwwojak replied:
Will resolve pending commit.

ecormany commented on api/docs/v2/new_labware.rst:
```diff
@@ -113,7 +170,7 @@ There are many different ways to access wells inside labware. Different methods
 +-------------------------------------+-------------------------------------------------------------------------------------------------------------------+
```
now might be a good opportunity to refactor this into a 3-column list-table.

jwwojak replied:
Trying a table. The `code formatting` takes a lot of space. Text is going to break across a line.

2 columns might be fine. 2nd column: definition, line break and then example.

Here's the revised 3 col table

```
.. list-table::
   :widths: 20 30 50
   :header-rows: 1

   * - Method
     - Returns
     - Example
   * - :py:meth:`.Labware.wells`
     - List of all wells.
     - ``[labware:A1, labware:B1, labware:C1...]``
   * - :py:meth:`.Labware.rows`
     - List of a list ordered by row.
     - ``[[labware:A1, labware:A2...], [labware:B1, labware:B2..]]``
   * - :py:meth:`.Labware.columns`
     - List of a list ordered by column.
     - ``[[labware:A1, labware:B1..], [labware:A2, labware:B2..]]``
   * - :py:meth:`.Labware.wells_by_name`
     - Dictionary with well names as keys.
     - ``{'A1': labware:A1, 'B1': labware:B1}``
   * - :py:meth:`.Labware.rows_by_name`
     - Dictionary with row names as keys.
     - ``{'A': [labware:A1, labware:A2..], 'B': [labware:B1, labware:B2]}``
   * - :py:meth:`.Labware.columns_by_name`
     - Dictionary with column names as keys.
     - ``{'1': [labware:A1, labware:B1..], '2': [labware:A2, labware:B2..]}``
   ```

jwwojak replied:
Will resolve pending commit.

ecormany reviewed (changes requested):
A few mechanical (English and RST) comments, and a few structural comments. Biggest change will be getting `load_adapter()` fully documented in the Loading Labware on Adapters section.

jwwojak committed:
Update api/docs/v2/new_labware.rst

Will commit and resolve.

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

jwwojak committed:
Apply suggestions from code review

Batch commit of suggestions.

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

jwwojak reviewed (commented):
I think I've got all the corrections and changes. Made some locally and will commit to the PR.

jwwojak committed:
Commit with review changes

Made changes from reviewer feedback.

jwwojak committed:
Remove commented text

Had some commented text. Removing that.

jwwojak committed:
Note text and table fix

Adding version 2.14 or earlier text to the top of page note.

Fixed/added ... to the labware types table.

ecormany committed:
proofread pass

ecormany committed:
rework adapter §

ecormany committed:
prose and code cleanup

ecormany reviewed (approved):
I give the 👍 on the prose. I mostly made style edits, although I rewrote the Loading Labware on Adapters section. @jwwojak, I'd appreciate your feedback there.

For code stuff, especially the new ``load_adapter()`` behavior, adding @jbleon95 as a reviewer and waiting for his 2¢ before merging.

jwwojak reviewed (commented):
@ecormany Regarding this text in Loading Labware (my italics):

> But you can also load labware on top of an adapter that either fits on a module or goes directly on the deck. _Compatible labware then sits atop the adapter_.

That second sentence seems redundant. The previous already suggests labware on top of an adapter. Can this be removed?

Also, word choice:  "atop"? I get it. This lets us avoid using "on top of" twice, but still something about it. It works but it reminds me of using a close matching SAE socket on a metric bolt.

But these are minor and shouldn't stop you from merging.

ecormany committed:
less redundant redundancy

ecormany committed:
labware-on-adapters anchor

jbleon95 reviewed (approved):
Adapter and labware loading onto adapters looks good.
